### PR TITLE
Threaded Example using New API

### LINKED ISSUE #221: Implement infrastructure for a channel-based example
The current networking example has a lot of code based around creating and using actual network channels from a server. Ideally, the primary demo for the library will focus on how to use the API and security-critical modules the calling application needs to implement (e.g. persistent storage).

## Proposed solution
@gatoWololo proposed creating an example that runs participants on different threads and sends messages over channels.

To keep this issue reasonably sized, the example should not aim to be fully-featured at this point (e.g. no user input to select protocols). It maybe should not even use `Participant`s yet; focus instead on generating threads and setting up channels.

## Completion criteria
- [ ] Make a new example with a good name
- [ ] Write a bit of initial documentation explaining the goals of the example
- [ ] Implement the infrastructure to set up different threads and communicate between them

### PR DESCRIPTION
**Pull request by gatoWololo**
Closes #221

Here is the threaded example.

Overall the public API is shaping out well. I like that we expliclity pass the necessary inputs to each subprotocol and get the values back out.

There is some rough edges that I will add as comments in this PR.

### CODE REVIEW

File: src/auxinfo/mod.rs
Comment (gatoWololo) [inline src/auxinfo/mod.rs:13]: The `status` associated type of implementors of `ProtocolParticipant` is not currently very useful. It is not exposed to users, so there is not much that I can do with them. I exposed them for this example, but ended up not actually using it.

File: src/lib.rs
Comment (gatoWololo) [inline src/lib.rs:109]: I expose these modules to access their respective `Status` enum

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:143]: Workers appear to return empty vectors of messages sometimes. We should wrap the `Vec<Message>` in an option instead of relying on the empty behavior, this way users don't have to check if the vec is empty. This is specially important when we will be sending these messages over the network.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:186]: Currently, it is possible to instantiate a standalone `KeygenParticipant` but I think the proper way to use our participants is always in the form of `Participant<P>` where `P: ProtocolParticipant`? We should remove the constructors that allow creating the "inner" type.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:187]: I'm not sure if I like the concept of a `config`, at least the name? But this is a minor nitpick.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:201]: I was surprised that aux info took no inputs? Is this correct?

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:223]: I know there is already a comment in the code and an issue for this.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:235]: Is this how I'm supposed to generate a `Sha256` to pass to our sign function? I wasn't sure.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:252]: It is a bit awkward to pass the rng to every call to process_single_message. The fact it needs a rng seems like an impl detail.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:262]: How useful do you think the `Status` associated type is? I see that often it is a 2-variant enum of either `Initialized` or `TerminatedSuccessfully`.
- Does every subprotocol need its own?

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:265]: Instead of using `status` I just end up checking if I have some output to see if we finished.

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:41]: I would also like your overall opinion on this example?

File: test_application/src/main.rs
Comment (gatoWololo) [inline test_application/src/main.rs:224]: I wasn't sure what to do with the computed `SignatureShare`s? do these need to be combined back together? Does the order you do it matter?

File: test_application/src/main.rs
Comment (marsella) [reply test_application/src/main.rs:41]: I like it! I had to take a couple of passes to understand the channel setup and determine which functions are run centrally vs on threads since I'm not familiar with that kind of infrastructure. It would help me to have a little bit of documentation on the functions themselves to indicate role + which party (worker vs router) executes them. 

I like that it's small and that it doesn't get bogged down in handling all the potential failure cases. One potential gap is in handling caller-responsible error cases -- `CallerError` variant that means the router messed up somehow, and instead of having the entire worker bork in that case, it would be nice to keep it alive and have it inform the router of the mistake. It's not necessary here, of course, since the example should always work because all the workers are honest, but it would be helpful I think as a guide to users. But I'm glad there isn't a bunch of code about network failures and other types of mistakes.

File: test_application/src/main.rs
Comment (marsella) [inline test_application/src/main.rs:29]: This is great naming. It actually never occurred to me that the end of a round is exactly the point at which a participant will have messages to send out.

**Review by marsella — COMMENTED** [review]
Thanks for this example, I really like it! I responded to your comments inline.

We talked about having an offline discussion about this to figure out next steps. I'll approve properly after that.

File: test_application/src/main.rs
Comment (marsella) [inline test_application/src/main.rs:250]: question: Do you like the shape of this API where you have to separately create the `Participant` and then initialize it? It feels a bit clunky to me in the current format to have to create this message and then pass it back the participant to process, but I'm also asking about the create / initialize separation (as opposed to having the `from_config` function return a bunch of messages to start sending out).

**Review by marsella — COMMENTED** [review]
One more question! I'm curious why this is a standalone binary instead of living in the `examples` directory.

**Comment by marsella** [comment]
Could you please link this to the issue it addresses?

File: test_application/src/main.rs
Comment (marsella) [reply test_application/src/main.rs:334]: I just looked more carefully at the `PresignRecord::sign` function and it actually calls `finalize` internally. I guess this was a design choice to force the caller to use a sha256 hash specifically, and not just a random 256 byte string.

File: test_application/src/main.rs
Comment (marsella) [reply test_application/src/main.rs:361]: I made #311 to try to make this more usable

**Comment by marsella** [comment]
Here are the issues I've identified so far.
- [ ] Statuses are not useful. I made #311 as a first step toward making this better.
- [ ] Return optional vectors from `process_single_message` instead of empty ones.
- [ ] Revisit protocol participant trait API and see if there's a way we can limit external access to `ProtocolParticipant` methods while keeping the nice generic properties we currently have.
- [ ] Make a plan for RNGs and participants. I want to think / play with the implementation details here to make sure we can require a reseedable RNG. I can write up this issue because I think I have more context on it; it might be better scoped as an epic but I'll try to get down the requirements for now.

File: src/auxinfo/mod.rs
Comment (gatoWololo) [reply src/auxinfo/mod.rs:13]: This is solved by this PR.

File: src/lib.rs
Comment (gatoWololo) [reply src/lib.rs:109]: Moved to #314

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs]: Moved to #316

**Comment by gatoWololo** [comment]
> One more question! I'm curious why this is a standalone binary instead of living in the examples directory.

I just changed it to using the `examples/` directory. Originally I thought I couldn't have multiple files for an example, but you can

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs:349]: I do not like it. I agree it is clunky and feels unnecessary. I will make an issue about this.

Comment (gatoWololo) [reply test_application/src/main.rs:349]: See #317

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs:334]: Conversation moved to #251

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs:351]: Continuing conversation on #290

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs:169]: Makes sense, I would also like to see workers robust to failure and inform the main thread. We can fix this in the follow "networking" example.

File: test_application/src/main.rs
Comment (gatoWololo) [reply test_application/src/main.rs:204]: I moved it to a more appropriate place during my refactor! Lmk if the new place is good.

File: examples/test_application/src/main.rs
Comment (marsella) [inline examples/test_application/src/main.rs:67]: nit: did `cargo fmt` pull your list into one line?

File: examples/test_application/src/main.rs
Comment (marsella) [inline examples/test_application/src/main.rs:81]: suggestion: `StdRng` is somewhat faster than going straight to the OS for randomness and would be suitable for this setting. I think the `StdRng::from_entropy` constructor gets its seed from OsRng anyway.

File: examples/test_application/src/main.rs
Comment (marsella) [inline examples/test_application/src/main.rs:83]: nit: us -> use?

File: examples/test_application/src/main.rs
Comment (marsella) [inline examples/test_application/src/main.rs:4]: suggestion: maybe link to the mpsc page here for those who don't know what this is? https://doc.rust-lang.org/std/sync/mpsc/

File: examples/test_application/src/main.rs
Comment (marsella) [inline examples/test_application/src/main.rs:103]: thought: should we make the `random_quorum` function on `ParticipantConfig` public?

File: src/lib.rs
Comment (marsella) [inline src/lib.rs:155]: nit: remove outdated docs (I think you handled this?)

File: Cargo.toml
Comment (marsella) [inline Cargo.toml:1]: todo: I think this needs to be updated to explicitly include the example, since automatic example detection is turned off on L11.

```
[[example]]
name = "test_application"
path = "examples/test_application/src/main.rs"
```

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks Omar, this looks really good! I appreciate all the additional documentation.

The only blocking comment I have is about adding this to the `Cargo.toml`.

File: examples/test_application/src/main.rs
Comment (gatoWololo) [reply examples/test_application/src/main.rs:67]: Ugh, probably because I swear I fixed it.

File: examples/test_application/src/main.rs
Comment (gatoWololo) [reply examples/test_application/src/main.rs:103]: That makes sense! I added it to this PR and fixed the doc string.

File: src/lib.rs
Comment (marsella) [reply src/lib.rs:155]: Did you mean to leave this in? If so, could you reference #314, or whichever issue you think will address it?

**Review by marsella — APPROVED** [review]
Thanks, this looks good to me!

**Comment by marsella** [comment]
Flagging [comment](https://github.com/boltlabs-inc/tss-ecdsa/pull/302#discussion_r1182962625) that is just a nit but you may want to address before merging.

File: src/lib.rs
Comment (gatoWololo) [reply src/lib.rs:155]: Whoops. I thought I had addressed it, thanks!

**Pull request merged**
